### ISSUE TITLE
Most of Coroutine-related rules do not work

### ISSUE THREAD

**Issue by fraggjkee** [detekt/detekt#6714]
## Expected Behavior
Test class I'm using:
```kotlin
class DetektTest {

    // RedundantSuspendModifier
    suspend fun unneededSuspend() {
        Log.d("test", "test!")
    }

    // InjectDispatcher
    suspend fun nonInjectedDispatcher() {
        withContext(Dispatchers.IO) {
            delay(1)
        }
        withContext(Dispatchers.Main) {
            delay(2)
        }
    }

    // SleepInsteadOfDelay
    @Suppress("RedundantSuspendModifier")
    suspend fun foo() {
        Thread.sleep(1_000L)
    }

    // GlobalCoroutineUsage
    fun globalScope() {
        GlobalScope.launch {
            delay(100)
            Log.d("test", "globalScope")
        }
    }
}
```

It violates most of the rules from [Coroutines](https://detekt.dev/docs/rules/coroutines) section so I expect to get multiple errors generated by `./gradlew detekt`.

## Observed Behavior
Only 1 issue is reported: `GlobalCoroutineUsage `. 0 issues is reported if I remove `globalScope ` method.

```
astulnev@de-astulnev Detekttest % ./gradlew detekt

> Task :app:detekt FAILED
/Users/astulnev/Desktop/Detekttest/app/src/main/java/com/example/detekttest/MainActivity.kt:68:9: This use of GlobalScope should be replaced by `CoroutineScope` or `coroutineScope`. [GlobalCoroutineUsage]

FAILURE: Build failed with an exception.
```
![Screenshot 2023-12-13 at 13 33 26](https://github.com/detekt/detekt/assets/1222517/d3964441-0e16-4709-973d-f0e3e55822b6)

## Steps to Reproduce
- clone this[ demo project](https://github.com/fraggjkee/detekt-bugreport)
- run `./gradlew detekt`

## Your Environment
* Version of detekt used: 1.23.3
* Operating System and version: macOS Sonoma 14.1.2
* Android Project + Android Studio Hedgehog | 2023.1.1

Config:
```
build:
  maxIssues: 0
  excludeCorrectable: false
  weights:
  # complexity: 2
  # LongParameterList: 1
  # style: 1
  # comments: 1

config:
  validation: true
  warningsAsErrors: true
  checkExhaustiveness: false
  # when writing own rules with new properties, exclude the property path e.g.: 'my_rule_set,.*>.*>[my_property]'
  excludes: ''

processors:
  active: true
  exclude:
    - 'DetektProgressListener'
  # - 'KtFileCountProcessor'
  # - 'PackageCountProcessor'
  # - 'ClassCountProcessor'
  # - 'FunctionCountProcessor'
  # - 'PropertyCountProcessor'
  # - 'ProjectComplexityProcessor'
  # - 'ProjectCognitiveComplexityProcessor'
  # - 'ProjectLLOCProcessor'
  # - 'ProjectCLOCProcessor'
  # - 'ProjectLOCProcessor'
  # - 'ProjectSLOCProcessor'
  # - 'LicenseHeaderLoaderExtension'

console-reports:
  active: true
  exclude:
    - 'ProjectStatisticsReport'
    - 'ComplexityReport'
    - 'NotificationReport'
    - 'FindingsReport'
    - 'FileBasedFindingsReport'
  #  - 'LiteFindingsReport'

output-reports:
  active: true
  exclude:
    - 'TxtOutputReport'
    - 'XmlOutputReport'
    - 'MdOutputReport'
    - 'SarifOutputReport'
  # - 'HtmlOutputReport'

coroutines:
  active: true
  GlobalCoroutineUsage:
    active: true
  InjectDispatcher:
    active: true
    dispatcherNames:
      - 'IO'
      - 'Default'
      - 'Unconfined'
  RedundantSuspendModifier:
    active: true
  SleepInsteadOfDelay:
    active: true
  SuspendFunSwallowedCancellation:
    active: true
  SuspendFunWithCoroutineScopeReceiver:
    active: true
  SuspendFunWithFlowReturnType:
    active: true
```

Gradle setup:
```
detekt {
    toolVersion = "1.23.3"
    config.setFrom(file("detekt.yml"))
    buildUponDefaultConfig = false
}
```

**Comment by fraggjkee** [comment]
This seems not to be specific to Coroutines test suite, I found a bunch of other rules that don't report corresponding issues, even when those issues are copy-pasted from the Wiki as a "Non-compiant" examples, a quick example is https://detekt.dev/docs/rules/potential-bugs#ignoredreturnvalue that does not fail for cases like

```
fun flow(): Flow<String> = emptyFlow()

fun otherFunc() {
    flow() // ignoring result

    flowOf(1, 2, 3) // also ignoring result
}
```

Is something wrong with my setup or many of the rules are really non-functional?

**Comment by fraggjkee** [comment]
Closing as the issue was indeed related to non-obvious detail with Type Resolution, wasn't quite expecting most of the rules to be ignored by the default `detekt` task :-( 

Trying to find a way to properly setup type resolution-enabled Detekt in a multi-module Android project,  does not seem very straightforward so far as I haven't found anything related to report merging or `checkDependencies` from Android lint that allows to just run the check against the main app module visiting all the children transitively.
